libs/vec/native/src/vec/c/vec.c

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+#include <arm_neon.h>
+#include "vec.h"
+
+EXPORT int stride() {
+    return STRIDE_BYTES_LEN;
+}
+
+EXPORT int dot8s(const void* a, const void* b, int dims) {
+    // We have contention in the instruction pipeline on the accumulation
+    // registers if we use too few.
+    int32x4_t acc1 = vdupq_n_s32(0);
+    int32x4_t acc2 = vdupq_n_s32(0);
+    int32x4_t acc3 = vdupq_n_s32(0);
+    int32x4_t acc4 = vdupq_n_s32(0);
+
+    // Some unrolling gives around 50% performance improvement.
```

Comment:
I got a benefit from unrolling this with clang. I guess we either build with clang or gcc. I would at least test with this:
```
    #ifdef __clang__
    #pragma clang loop unroll_count(2)
    #elif __GNUC__
    #pragma GCC unroll 2
    #endif
```

Reply:
This is hand unrolled, so striding 32 bytes at a time. Are you suggesting to further unroll, so 64 bytes at a time? Or to remove the hand unrolling and leave it to the compiler?     I avoided the latter, as the Java code needs to know the stride, so that it can calculate the tail correctly. I'm not sure that the former offers us much additional benefit.

Reply:
I found the combination of hand unrolling (the main benefit being you accumulate into multiple registers so the four different load, multiply and accumulate instructions can be pipelined with no waiting) PLUS this compiler directive gained a bit more. It was around +4% for adding the compiler directive on my M2.

There is no way to replicate the multiple accumulators with compiler directives and this effect is large. Conversely if you use the compiler directive you aren't obliged to have your loop a multiple of 64, so you are less likely to run into needing a remainder loop. 

So I do really think there is a small advantage in the combination. Although 4% may not matter too much and may vary according to the actual chip you're running on, so I'm also happy if you omit.